cryptoki/src/mechanism/hkdf.rs

```diff
@@ -0,0 +1,116 @@
+//! Mechanisms of hash-based key derive function (HKDF)
+//! See: <https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061597>
+
+use std::{convert::TryInto, marker::PhantomData, ptr::null_mut, slice};
+
+use cryptoki_sys::{CKF_HKDF_SALT_DATA, CKF_HKDF_SALT_KEY, CKF_HKDF_SALT_NULL};
+
+use crate::object::ObjectHandle;
+
+use super::MechanismType;
+
+#[derive(Debug, Clone, Copy)]
+/// The salt for the extract stage.
+pub enum HkdfSalt<'a> {
+    /// CKF_HKDF_SALT_NULL no salt is supplied.
+    Null,
+    /// CKF_HKDF_SALT_DATA salt is supplied as a data in pSalt with length ulSaltLen.
+    Data(&'a [u8]),
+    /// CKF_HKDF_SALT_KEY salt is supplied as a key in hSaltKey
+    Key(ObjectHandle),
+}
+
+/// HKDF parameters.
+///
+/// This structure wraps a `CK_HKDF_PARAMS` structure.
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct HkdfParams<'a> {
+    inner: cryptoki_sys::CK_HKDF_PARAMS,
+    /// Marker type to ensure we don't outlive the data
+    _marker: PhantomData<&'a [u8]>,
+}
+
+impl<'a> HkdfParams<'a> {
+    /// Construct parameters for hash-based key derive function (HKDF).
+    ///
+    /// # Arguments
+    ///
+    /// * `extract` - Whether to execute the extract portion of HKDF.
+    ///
+    /// * `expand` - Whether to execute the expand portion of HKDF.
+    ///
+    /// * `prf_hash_mechanism` - The base hash used for the HMAC in the underlying HKDF operation
+    ///
+    /// * `salt` - The salt for the extract stage.
+    ///
+    /// * `info` - The info string for the expand stage.
+    pub fn new(
+        extract: bool,
+        expand: bool,
+        prf_hash_mechanism: MechanismType,
+        salt: HkdfSalt,
+        info: &'a [u8],
+    ) -> Self {
+        Self {
+            inner: cryptoki_sys::CK_HKDF_PARAMS {
+                bExtract: extract as u8,
+                bExpand: expand as u8,
+                prfHashMechanism: *prf_hash_mechanism,
+                ulSaltType: match salt {
+                    HkdfSalt::Null => CKF_HKDF_SALT_NULL,
+                    HkdfSalt::Data(_) => CKF_HKDF_SALT_DATA,
+                    HkdfSalt::Key(_) => CKF_HKDF_SALT_KEY,
+                },
+                pSalt: match salt {
+                    HkdfSalt::Data(data) => data.as_ptr() as *mut _,
+                    _ => null_mut(),
+                },
+                ulSaltLen: match salt {
+                    HkdfSalt::Data(data) => data
+                        .len()
+                        .try_into()
+                        .expect("salt length does not fit in CK_ULONG"),
+                    _ => 0,
+                },
+                hSaltKey: match salt {
```

Comment:
I'd convert all these two-arm matches to `if let`s... but this is purely a stylistic choice :sweat_smile: